Sming/Libraries/OtaUpgrade/OtaUpgrade/SignatureVerifier.h

```diff
@@ -0,0 +1,58 @@
+/****
+ * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
+ * Created 2017 by Slavey Karadzhov
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * OtaSignatureVerifier.h
+ *
+ ****/
+
+#pragma once
+
+#include <sodium/crypto_sign.h>
+#include <FlashString/Array.hpp>
+#include <array>
+
+namespace OtaUpgrade
+{
+DECLARE_FSTR_ARRAY(SignatureVerificationKey, uint8_t)
+
+/**
+ * @brief Signature verifier for `BasicStream`.
+ *
+ * This is a simple C++ wrapper for Libsodium's 
+ * <a href="https://download.libsodium.org/doc/public-key_cryptography/public-key_signatures">`crypto_sign_...`</a> API.
+ */
+class SignatureVerifier
+{
+public:
+	typedef std::array<uint8_t, crypto_sign_BYTES> VerificationData; ///< Signature type
+
+	SignatureVerifier()
+	{
+		crypto_sign_init(&state);
+	}
+
+	/** Continue incremental calculation for given chunk of data.
+	 */
+	void update(const void* data, size_t size)
+	{
+		crypto_sign_update(&state, static_cast<const unsigned char*>(data), size);
+	}
+
+	/** Check if \a signature matches the data previously fed into #update() calls.
+	 * @return `true` if \a signature matches data, `false` otherwise.
+	 */
+	bool verify(const VerificationData& signature)
+	{
+		assert(SignatureVerificationKey.length() == crypto_sign_PUBLICKEYBYTES);
+		LOAD_FSTR_ARRAY(verificationKey, SignatureVerificationKey);
+		return (crypto_sign_final_verify(&state, signature.data(), verificationKey) == 0);
```

Comment:
You could use SHA512 from crypto, then call `_crypto_sign_ed25519_verify_detached` directly...

Reply:
I note all the tables in libsodium are in RAM...

Reply:
Forget it, I notice _crypto_sign_ed25519_verify_detached  calls sha512 again. Is it a candidate for pulling into crypto, perhaps?

Reply:
I've pushed a possible solution to this to your repo...

Reply:
> I note all the tables in libsodium are in RAM...

That should probably be fixed all over libsodium. I can address this with a separate PR, have to change something in there anyway.

Regarding the `SignEd25519` class: On the one hand I would like an OOP interface to all of the crypto routines. On the other hand, I don't like that there are now two implementations of the same algorithm within the library. This also makes it harder to upgrade the library in the future. Have to think about that.

Reply:
I think it's fine to port each routine as the need arises, probably easier than trying to predict usage. This OOP interface is minimal, libsodium still does the difficult stuff. Could remove the C code? Also think it needs its own test application.

Reply:
Guess you are right. Let's incrementally add thin OOP wrappers when they come in handy. 
I'd still keep the C functions for now. Maybe we can decorate them with a `SMING_DEPRECATED` to kindly guide users to the C++ interface.

Unfortunately, due to time limitations, I've to postpone further work on this PR till Friday, so please be patient.

Reply:
I've merged your commits with a minor renaming (`PrivateKey` -> `PublicKey` to avoid confusion).

I think further libsodium improvements deserve their own PR. I am also strongly in favor of a test suite. Maybe we can reuse some of libsodium's own tests. The [one](https://raw.githubusercontent.com/jedisct1/libsodium/master/test/default/sign.c) for the `crypto_sign_...` API is quite big, but maybe we don't have to test hundreds of messages like they do.

Reply:
@aemseemann Do you want me to have an initial pop at the test suite? I'd start with a copy of the SmingTest sample.